text/2018-02-20-crossbeam-atomic.md

```diff
@@ -0,0 +1,576 @@
+and [`AtomicInteger`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html)
+in Java.
+
+```rust
+union AtomicCell<T> {
+    // Use this when `size_of::<T>() <= size_of::<usize>()`.
+    //
+    // For example, in order to store a `val` into `atomic`, do:
+    //
+    // ```
+    // let mut value = 0usize;
+    // ptr::write_unaligned(&mut value as *mut T, val);
+    // self.atomic.store(value, ordering);
+    // ```
+    atomic: AtomicUsize,
+
+    // Use this when `size_of::<T>() > size_of::<usize>()`.
+    //
+    // Note that accessing this value requires locking a global spinlock.
+    // For example, in order to store a `val` into `cell`, do:
+    //
+    // ```
+    // // We're using the same `LOCKS` array from the `Atomic` API.
+    // let index = (&self as *const _ as usize) % NUM_LOCKS;
+    // while LOCKS[index].compare_and_swap(false, true, Acquire) {
+    //     ::std::thread::yield_now();
+    // }
+    // *(self.cell.get()) = val;
+    // LOCKS[index].store(false, Release);
+    // ```
+    cell: UnsafeCell<T>,
+}
+
+impl<T> AtomicCell<T> {
+    fn new(val: T) -> AtomicCell<T>;
+    fn get_mut(&mut self) -> &mut T;
+    fn into_inner(self) -> T
+
+    // Returns `true` when `size_of::<T>() <= size_of::<usize>()`.
+    fn is_lock_free() -> bool;
+
+    fn set(&self, val: T);
+    fn replace(&self, val: T) -> T;
+}
+
+impl<T: Default> AtomicCell<T> {
+    fn take(&self) -> T;
+}
+
+impl<T: Copy> AtomicCell<T> {
+    fn get(&self) -> T;
+
+    // We might want to restrict this method to primitive types rather than
+    // accepting any `T: Copy`.
+    fn compare_and_swap(&self, current: T, new: T) -> T;
+}
+
+// That's a lot of methods - maybe we should select just a subset of them.
+impl AtomicCell<bool> {
+    fn and(&self, val: bool);
+    fn nand(&self, val: bool);
+    fn or(&self, val: bool);
+    fn xor(&self, val: bool);
+
+    fn and_get(&self, val: bool) -> bool;
+    fn nand_get(&self, val: bool) -> bool;
+    fn or_get(&self, val: bool) -> bool;
+    fn xor_get(&self, val: bool) -> bool;
+
+    fn get_and(&self, val: bool) -> bool;
+    fn get_nand(&self, val: bool) -> bool;
+    fn get_or(&self, val: bool) -> bool;
+    fn get_xor(&self, val: bool) -> bool;
+}
+
+// Repeat this for `u8`, `i16`, `u16`, `i32`, `u32`, `i64`, `u64`, `isize`, `usize`.
+// That's a lot of methods - maybe we should select just a subset of them.
+impl AtomicCell<i8> {
+    fn add(&self, val: i8);
+    fn sub(&self, val: i8);
+    fn and(&self, val: i8);
+    fn or(&self, val: i8);
+    fn xor(&self, val: i8);
+
+    fn add_get(&self, val: i8) -> i8;
+    fn sub_get(&self, val: i8) -> i8;
+    fn and_get(&self, val: i8) -> i8;
+    fn or_get(&self, val: i8) -> i8;
+    fn xor_get(&self, val: i8) -> i8;
+
+    fn get_add(&self, val: i8) -> i8;
+    fn get_sub(&self, val: i8) -> i8;
+    fn get_and(&self, val: i8) -> i8;
+    fn get_or(&self, val: i8) -> i8;
+    fn get_xor(&self, val: i8) -> i8;
+}
```

Comment:
I don't think it's worth the trouble of adding these operations. You should just use `AtomicI8` or `Atomic<i8>` if you want atomic integer operations. These don't belong on a cell type.

Reply:
But `AtomicI8` and `Atomic<i8>` still require memory orderings on all operations. Those aren't designed for casual users of atomics.

We need *some* kind of primitive that has SeqCst-by-default atomic integer operations. That is one of the pain points from the motivation sections.